### PR TITLE
Replace jstree with jquery ui sortable

### PR DESCRIPTION
**Pull request by Sinetheta**
We were using [jsTree](https://www.jstree.com/) in this one place only, to draw a taxonomy tree and let people drag around taxon leafs. Since we already have jQuery UI sortable, I thought that we could eliminate a dependency.

## Changes

* Removed jsTree.
* Removed the rabl templates that it was using to fetch the taxonomy structure.
* Used an existing api endpoint `/api/taxonomies/2?set=nested` to render the entire tree.
* All levels of nesting are now always visible (are we worried about someone having 1000 taxons?).
* Edit and Delete are now buttons on each taxon instead of a context menu.
* Same-page rename is now gone. I would be happy to reimplement that if we can agree how it should work. But I don't know of another place in the backend where we offer it.
* Rollback is gone. The new error functionality is "omg, fetch and redraw the tree".

## Before
screen shot 2015-12-04 at 2 41 43 pm

## After
screen shot 2015-12-04 at 2 41 35 pm

### CODE REVIEW

**Comment by cbrunsdon** [comment]
[insert pictures with explanation too, please. would be appreciated].

**Comment by Sinetheta** [comment]
@mandily @tvdeyen I'd love input on this page while I'm editing. The biggest concerns for me are around adding new taxons.

On the old page you would right click a taxon and "Add" a new one as the new last child. That taxon would be called "New node" and your focus would be automatically set to edit that name in-page.

My current implementation has only an "Add Taxon" button in the header. Clicking that button inserts a taxon called "New node" as the new last child of the root taxon. In order to rename any new taxons a user would have to "edit" that taxon by leaving the page.

That was the easiest thing to implement, but I'd be happy to entertain alternative interactions.

**Comment by jhawthorn** [comment]
Thanks Kevin. jstree was terrible to use, flakey to use, and made us keep around a lot of cruft (custom json routes). I am very happy to see it go.

**Comment by tvdeyen** [comment]
Fun fact: I also have a branch for removing jstree, but hadn't a chance to work on it any further. 

I use the [jquery nested sortable plugin](http://mjsarfatti.com/sandbox/nestedSortable/). A new dependency, yes, but a very small one, though. 

The reason why I choosed this plugin was: nesting. Jquery sortable does not support nesting; only sorting elements on the same level. But we want to be able to move a node into another one, right?

Regarding the context menu: I would remove it and add buttons besides each node to add a child node. 

We should also clean up the table columns. AwesomeNestedSet does not need a position column and nested sets in general work with parent, left and right columns. But the taxons table has all of them. 

Maybe I can free some time to publish my branch. Would love to here your feedback on this.

File: api/config/routes.rb [hunk 1]
```diff
@@ -91,14 +91,7 @@
     resources :states, only: [:index, :show]
 
     resources :taxonomies do
-      member do
-        get :jstree
-      end
-      resources :taxons do
-        member do
-          get :jstree
```
Comment (BenMorganIO) [inline api/config/routes.rb]: We should probably drop this into an extension in case anyone has an iOS or Android app relying on these endpoints.

**Comment by Sinetheta** [comment]
Oh, but I'm sneaky @tvdeyen 

```coffee
$('#taxonomy_tree ul').sortable
  connectWith: '#taxonomy_tree ul'
```

Bam! nested sortables.

File: api/config/routes.rb (hunk 1 above)
Comment (Sinetheta) [inline api/config/routes.rb]: :unamused:

**Comment by tvdeyen** [comment]
Ha! You are great @Sinetheta 
I don't have the time right now, to play with it. Will do tomorrow, but this is very neat!

**Comment by Mandily** [comment]
I'd like to see the original 'move' icon used instead of the 'sortable' icon you've used here. A bit more space between the two icons would be nice too, so we can minimize the chance of the user missing their target and clicking on the wrong icon. 

Did you/can you remove that error message when the user tries to drag a row to the top level? It would be better if we could just not allow them to do that in the first place. 

Otherwise looks good!

File: api/config/routes.rb (hunk 1 above)
Comment (BenMorganIO) [inline api/config/routes.rb]: Just saying, removing from the API might really upset anyone who wants to upgrade, but they're using this in their iOS or Android app. I know my previous employer [Kabuni](https://kabuni.com) is using it in their iOS app.

**Comment by tvdeyen** [comment]
I tested it and find it very hard to drag a node into another one. It's so hard to hit the trigger, that I thought 
at first it is not even possible. I tried to use `tolerance: "pointer"`, but this didn't help either. 
And even if one finally manages to move a node into another one, it is nearly impossible to move it out again. Not a great UX IMO.

I am aware, that another dependency is not pleasant, but swapping out `jstree` with `jquery.nestedSortable` is still a total win of `148kb` and a much better UX for the user. 
Please try the demo at http://mjsarfatti.com/sandbox/nestedSortable/ to feel the difference.

If you still want to go only with `jquery.ui.sortable` I'm fine with that, but we should make the handling much smoother then, otherwise it does not feel like a good change UX wise.
Maybe some css (tried raising `min-height` for `#taxonomy_tree ul`) will do the trick.

**Comment by Sinetheta** [comment]
Thanks for taking a look at it @tvdeyen. Can you explain which drag operations were problematic. Personally I found it at least as functional as jsTree was (dragging any tree structure is going to require playing around until you get the hang of the particular implementation).

One thing that _was_ giving me a problem, was trying to get something into the bottom most position of any list (it always wanted to be a child, not a sibling of the element above it), so I tweaked some settings and I think it feels more intuitive now.

I'd be happy to give nestedSortable a shot though, if this still isn't feeling good enough.

File: api/config/routes.rb (hunk 1 above)
Comment (cbrunsdon) [inline api/config/routes.rb]: I'm more a fan of killing it, as it wasn't ever actually published in the public API on the old docs site:
https://guides.spreecommerce.com/api/taxonomies.html

We can make a note in the release notes pointing to the commit if they want to pull that code into their own stores. I would consider that a better upgrade path than supporting it as an extension.

Though not removing the routes and just axing the javascript is still a good improvement... Maybe we do leave it till we just deprecate the entirety of the API.

Comment (cbrunsdon) [inline api/config/routes.rb]: @Sinetheta yea if you can at least pull those commits that remove the ruby side out into a separate PR it would be nice.

**Comment by tvdeyen** [comment]
Feels much better now. Great work, Kevin!

There are still smaller UI glitches, but the dragging is perfect. 

Great to see that we don't need any additional plugin for that feature. 

Regarding the routes: Deprecate and then remove with the next release.

**Comment by Sinetheta** [comment]
I've added the api routes back in @BenMorganIO @cbrunsdon, we can just rip them out when we address the API as a whole.

**Comment by jhawthorn** [comment]
:+1:

**Comment by cbrunsdon** [comment]
Yea, I'm good on this one too, thanks @Sinetheta

**Comment by athal7** [comment]
:+1:

**Comment by simo163** [comment]
I just noticed that when creating new taxons, they don't get the right parent_id, but taxonomy_id (it worked for me only for first taxonomies of sample data whose ids are the same).

I changed the js files that taxonomy_id is not really needed, otherwise I find it must work if we force to assign the parent_id in the [api](https://github.com/solidusio/solidus/blob/master/api/app/controllers/spree/api/taxons_controller.rb#L39) by removing the unless statement.

**Pull request merged**